### PR TITLE
Map as batch result

### PR DESCRIPTION
**Pull request by bbakerman**
See #18

The first part of this URL is via PR #24  so read.  Because of context we want that in the new Map batch load code.

This adds the ability to have a batch loader function that returns a Map of values instead of a List.  Nulls will be used to indicate missing values.

It suits the case where given a list of N keys you can return < N values.

### CODE REVIEW

**Comment by bbakerman** [comment]
@aschrijver - any chance you could have a look at this PR.  I value your input

File: README.md [hunk 1]
```diff
@@ -165,6 +165,70 @@ a list of user ids in one call.
...
+Often there is not a 1:1 mapping of your batch loaded keys to the values returned.
+
+For example, let's assume you want to load users from a database, you could probably use a query that looks like this:
+
+```sql
+  SELECT * FROM User WHERE id IN (keys)
+```
+ 
+ Given say 10 user id keys you might only get 7 results back.  This can be more naturally represented in a map
+ than in an order list of values when returning values from the batch loader function.
```
Comment (aschrijver) [inline README.md]: correction: _ordered_ list

File: README.md [hunk 2]
```diff
@@ -165,6 +165,70 @@ a list of user ids in one call.
...
+ 
+ You can use `org.dataloader.MapBatchLoader` for this purpose. 
+ 
+ When the map is processed by the `DataLoader` code, any keys that are missing in the map
+ will be replaced with null values.  The semantics that the number of `DataLoader.load` requests
+ are matched with values is kept.
```
Comment (aschrijver) [inline README.md]: improvement: are matched with _an equal number of_ values is kept.

File: src/main/java/org/dataloader/BatchLoader.java [hunk 3]
```diff
@@ -74,11 +74,33 @@
...
      *
      * @param keys the collection of keys to load
      *
      * @return a promise of the values for those keys
      */
     CompletionStage<List<V>> load(List<K> keys);
+
+    /**
+     * Called to batch load the provided keys and return a promise to a list of values.  This default
+     * version can be given a context object to that maybe be useful during the call.  A typical use case
```
Comment (aschrijver) [inline src/main/java/org/dataloader/BatchLoader.java]: correction: can be given a context object _that may be_ useful during the call.

File: src/main/java/org/dataloader/BatchLoader.java [hunk 4]
```diff
@@ -74,11 +74,33 @@
...
+     * is passing in security credentials or database connection details say.
```
Comment (aschrijver) [inline src/main/java/org/dataloader/BatchLoader.java]: improvement: or database details, _for example._

File: src/main/java/org/dataloader/DataLoader.java [hunk 5]
```diff
@@ -134,6 +165,43 @@
         return new DataLoader<>((BatchLoader<K, V>) batchLoadFunction, options);
     }
 
+    /**
+     * Creates new DataLoader with the specified map abatch loader function and default options
```
Comment (aschrijver) [inline src/main/java/org/dataloader/DataLoader.java]: correction: map _batch_ loader function

File: src/main/java/org/dataloader/DataLoader.java [hunk 6]
```diff
@@ -134,6 +165,43 @@
...
+     * (batching, caching and unlimited batch size) where the batch loader function returns a list of
+     * {@link org.dataloader.Try} objects.
+     *
+     * This allows you to capture both the value that might be returned and also whether exception that might have occurred getting that individual value.  If its important you to
+     * know gather exact status of each item in a batch call and whether it threw exceptions when fetched then
```
Comment (aschrijver) [inline src/main/java/org/dataloader/DataLoader.java]: correction: line width, plus the first sentence is formulated confusingly.

File: src/main/java/org/dataloader/MapBatchLoader.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+import java.util.concurrent.CompletionStage;
+
+/**
+ * A function that is invoked for batch loading a map of of data values indicated by the provided list of keys. The
+ * function returns a promise of a map of results of individual load requests.
+ *
+ * There are a few constraints that must be upheld:
+ * <ul>
+ * <li>The keys MUST be able to be first class keys in a Java map.  Get your equals() and hashCode() methods in order</li>
+ * <li>The caller of the {@link org.dataloader.DataLoader} that uses this batch loader function MUSt be able to cope with
```
Comment (aschrijver) [inline src/main/java/org/dataloader/MapBatchLoader.java]: correction: _MUST_

File: src/main/java/org/dataloader/MapBatchLoader.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+ * than in an order list of values when returning values from the batch loader function.
+ *
+ * When the map is processed by the {@link org.dataloader.DataLoader} code, any keys that are missing in the map
+ * will be replaced with null values.  The semantics that the number of {@link org.dataloader.DataLoader#load(Object)} requests
+ * are matched with values is kept.
+ *
+ * This means that if 10 keys are asked for then {@link DataLoader#dispatch()} will return a promise 10 value results and each
+ * of the {@link org.dataloader.DataLoader#load(Object)} will complete with a value, null or an exception.
+ *
+ * When caching is disabled, its possible for the same key to be presented in the list of keys more than once.  You map
```
Comment (aschrijver) [inline src/main/java/org/dataloader/MapBatchLoader.java]: correction: _Your_ map

File: src/main/java/org/dataloader/MapBatchLoader.java [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+ * @param <V> type parameter indicating the type of values returned
+ *
+ * @author <a href="https://github.com/aschrijver/">Arnold Schrijver</a>
+ * @author <a href="https://github.com/bbakerman/">Brad Baker</a>
+ */
+@FunctionalInterface
+public interface MapBatchLoader<K, V> {
+
+    /**
+     * Called to batch load the provided keys and return a promise to a map of values.  It can be given a context object to
```
Comment (aschrijver) [inline src/main/java/org/dataloader/MapBatchLoader.java]: correction: context object _that may be_ useful

File: src/test/java/org/dataloader/DataLoaderContextTest.java [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.stream.Collectors;
+
+import static java.util.Arrays.asList;
+import static org.hamcrest.Matchers.equalTo;
+import static org.junit.Assert.assertThat;
+
+/**
+ * Tests related to context.  DataLoaderTest is getting to big and needs refactoring
```
Comment (aschrijver) [inline src/test/java/org/dataloader/DataLoaderContextTest.java]: correction: _too_ big
Might add a TODO to this sentence

File: src/test/java/org/dataloader/DataLoaderContextTest.java [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+
+        assertThat(results, equalTo(asList("A-ctx", "B-ctx", "C-ctx", "D-ctx")));
+    }
+
+    @Test
+    public void null_is_passed_as_context_if_you_do_nothing() throws Exception {
+        BatchLoader<String, String> batchLoader = new BatchLoader<String, String>() {
+            @Override
+            public CompletionStage<List<String>> load(List<String> keys) {
+                throw new UnsupportedOperationException("this wont be called");
```
Comment (aschrijver) [inline src/test/java/org/dataloader/DataLoaderContextTest.java]: correction: _won't_ be called

File: src/test/java/org/dataloader/DataLoaderMapBatchLoaderTest.java [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+import static org.dataloader.TestKit.futureError;
+import static org.dataloader.TestKit.listFrom;
+import static org.dataloader.impl.CompletableFutureKit.cause;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.instanceOf;
+import static org.hamcrest.Matchers.is;
+import static org.junit.Assert.assertThat;
+
+/**
+ * Much of the tests that related to {@link org.dataloader.MapBatchLoader} also related to
```
Comment (aschrijver) [inline src/test/java/org/dataloader/DataLoaderMapBatchLoaderTest.java]: correction: _are_ also related

**Comment by aschrijver** [comment]
Hi @bbakerman 

This PR looks quite good to me. Clean code, good documentation, tests. I didn't give it a spin, because I am not using graphql at the moment :( But I gave the documentation a good look, and marked some minor spelling errors and possible improvements here and there.

Regarding Javadocs I am wondering: You use list html and such, but I think any paragraph but the first should also have &lt;p&gt; tags, for HTML generation to work in all generators.

Nice work! It is great to see this evolving further and further :)

**Pull request merged**